hydrolib/core/io/ini/serializer.py

```diff
@@ -0,0 +1,246 @@
+    def from_section(cls, section: Section):
+        properties = list(
+            prop for prop in section.content if isinstance(prop, Property)
+        )
+
+        if len(properties) > 0:
+            max_key_length = max(len(prop.key) for prop in properties)
+            max_value_length = max(
+                len(prop.value) if prop.value is not None else 0 for prop in properties
+            )
+        else:
+            max_key_length = 0
+            max_value_length = 0
+
+        return cls(
+            max_key_length=max_key_length,
+            max_value_length=max_value_length,
+            max_datablock_element_length=_Lengths._calculate_max_length_datablock(
+                section.datablock
+            ),
+        )
+
+
+class Serializer:
+    """Serializer serializes Document to its corresponding lines."""
+
+    def __init__(self, config: SerializerConfig):
+        """Creates a new Serializer with the provided configuration.
+
+        Args:
+            config (SerializerConfig): The configuration of this Serializer.
+        """
+        self._config = config
+
+    @staticmethod
+    def _interweave(iterable: Iterable, val: Any) -> Iterable:
+        # Interweave the provided iterable with the provided value:
+        # iterable_element, val, iterable_element, val, ...
+
+        # Note that this will interweave with val without making copies
+        # as such it is the same object being interweaved.
+        return chain.from_iterable(zip(iterable, repeat(val)))
+
+    def _serialize_comment_block(
+        self, block: CommentBlock, indent_size: int = 0
+    ) -> Iterable[str]:
+        indent = " " * indent_size
+        return (f"{indent}{self._config.comment_delimiter} {l}" for l in block.lines)
+
+    def _serialize_comment_header(
+        self, header_blocks: Iterable[CommentBlock]
+    ) -> Iterable[str]:
+        blocks_as_lines = (
+            self._serialize_comment_block(block) for block in header_blocks
+        )
+        # We separate comment blocks in the header with an empty line.
+        blocks_as_lines = Serializer._interweave(blocks_as_lines, [""])  # type: ignore
+
+        return chain.from_iterable(blocks_as_lines)
+
+    def _serialize_section_header(self, section_header: str) -> Iterable[str]:
+        indent = " " * (self._config.section_indent)
+        return [f"{indent}[{section_header}]"]
+
+    @staticmethod
+    def _get_offset(key: Optional[str], max_length: int) -> str:
+        key_length = len(key) if key is not None else 0
+        return " " * max(max_length - key_length, 0)
+
+    def _serialize_property(
+        self, property: Property, lengths: _Lengths
+    ) -> Iterable[str]:
+        indent = " " * (self._config.total_property_indent)
+        key_offset = Serializer._get_offset(property.key, lengths.max_key_length)
+        key = f"{property.key}{key_offset} = "
+
+        value_offset = Serializer._get_offset(property.value, lengths.max_value_length)
```

Comment:
Isn't it technically 'value_offset' and 'comment_offset'? respectively

Reply:
I was thinking it is the offset associated with the key, as it is dependent on the length of the longest key within the section, thus key_offset made sense. Maybe key_whitespace is a better term though. I will change that.